### DISCUSSION TITLE
How to perform a deep merge of secrets in ecs services, like how it's done for environment

### DISCUSSION THREAD

**Discussion by umm0n** [gruntwork-io/knowledge-base#578] [Help]
When deploying the ref arch and using the envcommon <> child module configuration for ecs service, envcommon can be used to set global envars and the child module can be used to set more specific ones. On apply, a deep merge is performed, so that both values are saved.

I'm talking about this:

Locals
```
  service_environment_variables = {
    (include.envcommon.locals.service_name) = [
      {
        name  = "USUAL_ENV"
        value = "this-usual-env"
      }
    ],
    "fluentbit"     = []
    secrets         = []
  }

```
Inputs
```
  container_definitions = [
  for name, definition in include.envcommon.inputs._container_definitions_map :
  merge(
    definition,
    {
      name        = name
      image       = local.container_images[name]
      environment = concat(definition.environment, local.service_environment_variables[name])
    },
  )

  ]
```

Can the same be done for secrets, so that some are defined globally in envcommon and other, more specific ones, in the individual modules themselves? I haven't had luck trying to replicate it.

[r:terraform-aws-ecs](https://github.com/gruntwork-io/terraform-aws-ecs)
---

<ins datetime="2022-10-26T21:48:09Z">
  <p><a href="https://support.gruntwork.io/hc/requests/109450">Tracked in ticket #109450</a></p>
</ins>

**Comment by josh-padnick** [comment]
We generally recommend passing in an AWS Secrets Manager ARN as ECS service environment variables. Then within your service, use AWS API calls to AWS Secrets Manager to decrypt the appropriate secret. 

Based on this:

> When deploying the ref arch and using the envcommon <> child module configuration for ecs service, envcommon can be used to set global envars and the child module can be used to set more specific ones. On apply, a deep merge is performed, so that both values are saved. 

It sounds like you'd be able to use both "global" and "local" variables as desired to achieve this. Does that address your question?
